data/src/main/java/com/microsoft/azure/kusto/data/ClientImpl.java

```diff
@@ -194,43 +267,36 @@ public String executeToJsonResult(String database, String command) throws DataSe
-        String clusterEndpoint = String.format(commandType.getEndpoint(), clusterUrl);
+        KustoRequest kr = new KustoRequest(command, database, properties);
+        return executeToJsonResult(kr);
+    }
 
-        Map<String, String> headers;
+    private String executeToJsonResult(KustoRequest kr) throws DataServiceException, DataClientException {
 
-        try {
-            headers = generateIngestAndCommandHeaders(properties, "KJC.execute",
-                    commandType.getActivityTypeSuffix());
-            validateEndpoint();
-        } catch (KustoClientInvalidConnectionStringException e) {
-            throw new DataClientException(clusterUrl, e.getMessage(), e);
-        }
-        addCommandHeaders(headers);
-        String jsonPayload = generateCommandPayload(database, command, properties);
-        StringEntity requestEntity = new StringEntity(jsonPayload, ContentType.APPLICATION_JSON);
-        // trace execution
+        KustoRequestContext request = prepareRequest(kr);
+
+        // Get the response and trace the call
+        return MonitoredActivity.invoke(
+                (SupplierOneException<String, DataServiceException>) () -> post(request.getHttpRequest()),
+                request.getSdkRequest().getCommandType().getActivityTypeSuffix().concat(".executeToJsonResult"));
+    }
+
+    public Mono<String> processRequestAsync(KustoRequestContext request) {
         return MonitoredActivity.invoke(
-                (SupplierTwoExceptions<String, DataServiceException, DataClientException>) () -> HttpPostUtils.post(httpClient, clusterEndpoint, requestEntity,
-                        timeoutMs + CLIENT_SERVER_DELTA_IN_MILLISECS, headers),
-                commandType.getActivityTypeSuffix().concat(".executeToJsonResult"));
+                (SupplierNoException<Mono<String>>) () -> postAsync(request.getHttpRequest())
+                        .onErrorContinue((err, src) -> LOGGER.error("Error coming from src {}", src, err)),
+                request.getSdkRequest().getCommandType().getActivityTypeSuffix().concat(".executeToJsonResult"));
     }
 
-    private void validateEndpoint() throws DataServiceException, KustoClientInvalidConnectionStringException {
-        if (!endpointValidated) {
-            KustoTrustedEndpoints.validateTrustedEndpoint(clusterUrl,
-                    CloudInfo.retrieveCloudInfoForCluster(clusterUrl).getLoginEndpoint());
-            endpointValidated = true;
+    private void validateEndpoint() throws DataServiceException, DataClientException {
+        try {
+            if (!endpointValidated) {
+                KustoTrustedEndpoints.validateTrustedEndpoint(clusterUrl,
+                        CloudInfo.retrieveCloudInfoForCluster(clusterUrl).getLoginEndpoint());
+                endpointValidated = true;
+            }
+        } catch (KustoClientInvalidConnectionStringException e) {
+            throw new DataClientException(clusterUrl, e.getMessage(), e);
```

Comment:
Maybe we want an inhereted type of exception, so users could catch it specficially?

Reply:
I thought about this comment for a day or two and came to the conclusion that if we extend DataClientException or DataServiceException for most exception cases:

1. We don't have to touch method signatures when throwing DataClient or DataService
2. We don't have to catch the exceptions that extend DataClient or DataService and rethrow them
3. Customers can then catch more specific exceptions if that's something they want, and then do with that additional context what they will